src/main/kotlin/graphics/scenery/DFTParser.kt

```diff
@@ -0,0 +1,132 @@
+                        unitCellDimensions[2] = (gridSpacings[2]*gridDimensions[2])+unitCellOrigin[2]
+                        electronicDensity = Array(gridDimensions[0], { Array(gridDimensions[1],
+                            { FloatArray(gridDimensions[2]) } ) } )
+                    }
+                    // Parsing atomic positions.
+                    if (counter < 6+numberOfAtoms){
+                        atomicPositions[counter-6] = Vector3f((line.trim().split("\\s+".toRegex())[2]).toFloat(),
+                                                              (line.trim().split("\\s+".toRegex())[3]).toFloat(),
+                                                              (line.trim().split("\\s+".toRegex())[4]).toFloat())
+                    }
+
+                    // Parsing volumetric data.
+                    // A possible optimization here would be to read this into a 1D array. We cannot directly
+                    // read it into the byte buffer, because we don't know max/min values a-priori.
+                    if (counter >= 6+numberOfAtoms) {
+                        for (value in (line.trim().split("\\s+".toRegex()))) {
+                            // Cube files should be in Fortran (z-fastest ordering).
+                            electronicDensity[xcounter][ycounter][zcounter] = (value).toFloat()
+                            if (electronicDensity[xcounter][ycounter][zcounter] > maxDensity) {
+                                maxDensity = electronicDensity[xcounter][ycounter][zcounter]
+                            }
+                            if (electronicDensity[xcounter][ycounter][zcounter] < minDensity) {
+                                minDensity = electronicDensity[xcounter][ycounter][zcounter]
+                            }
+
+                            zcounter++
+                            if (zcounter == gridDimensions[2]) {
+                                zcounter = 0
+                                ycounter++
+                                if (ycounter == gridDimensions[1]) {
+                                    ycounter = 0
+                                    xcounter++
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            counter++
+        }
+        // Converting to byte buffer.
+        counter = 0
+        electronicDensityUInt =  MemoryUtil.memAlloc((gridDimensions[0]*
+            gridDimensions[1]*gridDimensions[2]*1))
+        if (normalizeDensityTo > 0.0f){
+            maxDensity = normalizeDensityTo
+        }
+        for (z in 0 until gridDimensions[2]){
+            for (y in 0 until gridDimensions[1]){
+                for (x in 0 until gridDimensions[0]){
+                    val value = (((electronicDensity[x][y][z] - minDensity) / (maxDensity - minDensity)) * 255.0f).toInt()
+                    electronicDensityUInt.put(value.toByte())
```

Comment:
If `electronDensityUInt` is supposed to be exposed publicly (as it is currently), I'd suggest working on a `duplicate()` of it. Otherwise using `put` will always move around the `position` of the ByteBuffer and will lead to nasty issues if someone else is accessing it while you're writing 👍

Reply:
Oh and also, is it really UInt and not UShort or UByte?

Reply:
You're absolutely right, this is a Byte array. I have changed the name accordingly. I have also implemented the usage of a `duplicate` (I hope I understood correctly what you meant).